[16.0][MIG] shopinvader_product_attribute_set

sbejaoui opened the pull request:
* [x] depends on https://github.com/OCA/odoo-pim/pull/142
* [x] depends on https://github.com/OCA/odoo-pim/pull/171

sbejaoui committed:
[16.0][MIG] shopinvader_product_attribute_set

lmignon commented on shopinvader_product_attribute_set/schemas/product.py:
```diff
@@ -0,0 +1,28 @@
+
+class ProductProduct(ProductProductBase, extends=True):
+    attribute_set: ProductAttributeSet | None = None
+    attributes: dict[str, Any] = {}
+    structured_attributes: list[dict[str, Any]] = [{}]
```
Why `Any`? We must be more precise...

lmignon commented on shopinvader_product_attribute_set/schemas/product.py:
```diff
@@ -0,0 +1,28 @@
+
+    @classmethod
+    def from_product_product(cls, odoo_rec):
+        obj = super(ProductProduct, cls).from_product_product(odoo_rec)
```
```suggestion
        obj = super().from_product_product(odoo_rec)
```

sbejaoui replied:
I am getting a strange error if I remove the class and cls from super

```
    @classmethod
    def from_product_product(cls, odoo_rec):
>       obj = super().from_product_product(odoo_rec)
E       TypeError: super(type, obj): obj must be an instance or subtype of type
```

lmignon replied:
@sbejaoui It's because your schema module is not loaded when the module is loaded by odoo and the extendable registry not properly initialized into your tests... see https://github.com/shopinvader/odoo-shopinvader/pull/1450/commits/85ba911e6a75b8a13b360003f73708e69cec19e8

lmignon reviewed (changes requested):
Can you specify the schema for attributes and attribute_set plz?

lmignon committed:
[FIX] shopinvader_product_attribute_set: Ensure schemas are loaded at loading time

To ensure that the schemas classes are put into the registry of extendable models when the addon is loaded, the schema module must be imported at module root. Moreover, in the tests, since the schemas are extendable class, we must ensure that the schema registry is properly initialized and reset at end of the tests. When extending schemas, it's also a good practice to call the methods overriden in the module on the schema's base classes. This ensures that the override is taken into account by the registry

lmignon commented:
@sbejaoui Test requirements are no more required

lmignon committed:
[FIX] shopinvader_product_attribute_set: Fix multiselect

lmignon commented:
IMO we should refactor the schema to be more precise about the constraint on the value type of attribute depending of the type of the attribute. I've some idea but no time at this moment.

lmignon commented:
/ocabot merge patch

simahawk commented on shopinvader_product_attribute_set/tests/test_product.py:
```diff
@@ -97,6 +111,119 @@ def test_product_attributes_empty_select(self):
+                "technical_description": "foo",
             },
         )
+        self.maxDiff = None
```
```suggestion
```

simahawk commented on shopinvader_product_attribute_set/tests/test_product.py:
```diff
@@ -97,6 +111,119 @@ def test_product_attributes_empty_select(self):
+            res.get("attributes"),
+            {"select": "Value 1", "multiselect": ["Value 1", "Value 2"]},
+        )
+        self.maxDiff = None
```
```suggestion
```

simahawk reviewed (commented):
Is this ready?

simahawk commented on shopinvader_product_attribute_set/schemas/product_attribute.py:
```diff
@@ -0,0 +1,61 @@
+    select = "select"
+    multiselect = "multiselect"
+    boolean = "boolean"
+    integer = "integer"
```
float is missing here

simahawk replied:
Fixed it here https://github.com/shopinvader/odoo-shopinvader/pull/1491
Feel free to fix here or to merge the other one. 
I just needed a complete PR to test ;)

lmignon replied:
Thank you @simahawk !
